Add initial codemod tooling

RobinMalfait opened the pull request:
This PR adds some initial tooling for codemods. We are currently only interested in migrating CSS files, so we will be using PostCSS under the hood to do this. This PR also implements the "migrate `@apply`" codemod from #14412.

The usage will look like this:

```sh
npx @tailwindcss/upgrade
```

You can pass in CSS files to transform as arguments:

```sh
npx @tailwindcss/upgrade src/**/*.css
```

But, if none are provided, it will search for CSS files in the current directory and its subdirectories.

```
≈ tailwindcss v4.0.0-alpha.24

│ No files provided. Searching for CSS files in the current
│ directory and its subdirectories…

│ Migration complete. Verify the changes and commit them to
│ your repository.
```

The tooling also requires the Git repository to be in a clean state. This is a common convention to ensure that everything is undo-able. If we detect that the git repository is dirty, we will abort the migration.

```
≈ tailwindcss v4.0.0-alpha.24

│ Git directory is not clean. Please stash or commit your
│ changes before migrating.

│ You may use the `--force` flag to override this safety
│ check.
```


---

This PR alsoo adds CSS codemods for migrating existing `@apply` directives to the new version.

This PR has the ability to migrate the following cases:

---

In v4, the convention is to put the important modifier `!` at the end of the utility class instead of right before it. This makes it easier to reason about, especially when you are variants.

Input:
```css
.foo {
  @apply !flex flex-col! hover:!items-start items-center;
}
```

Output:
```css
.foo {
  @apply flex! flex-col! hover:items-start! items-center;
}
```


---

In v4 we don't support `!important` as a marker at the end of `@apply` directives. Instead, you can append the `!` to each utility class to make it `!important`.

Input:
```css
.foo {
  @apply flex flex-col !important;
}
```

Output:
```css
.foo {
  @apply flex! flex-col!;
}
```

RobinMalfait committed:
init `@tailwindcss/upgrade`

RobinMalfait committed:
add codemod tooling

+ add codemod for migrating `@apply`

RobinMalfait committed:
copy utils from `@tailwindcss/cli`

Copied some utils from `@tailwindcss/cli`. Initially referenced them,
but we also rely on some dependencies that we have to install.

One solution is to move it to an internal package and import it so that
tsup can inline everything. This also has the same 'we need the correct
dependencies' problem.

Another solution is to create a proper `@tailwindcss/cli-utils` (or
similar) package that we can import as a dependency. However, this one
will require us to actually publish the package, and give it a proper
name.

RobinMalfait committed:
migrate passed in files (or discover files)

RobinMalfait committed:
add `error`, `info`, `success`, and `warn` helpers

philipp-spiess commented on packages/@tailwindcss-upgrade/package.json:
```diff
@@ -0,0 +1,45 @@
+{
+  "name": "@tailwindcss/upgrade",
```
Nit: The PR description mentions `@tailwindcss/codemod` and this here uses `@tailwindcss/upgrade` 😅

RobinMalfait committed:
sync pnpm-lock.yaml

RobinMalfait committed:
add integration test for the `@apply` codemod

RobinMalfait committed:
safely check whether we are in a git directory

RobinMalfait committed:
use binary directly

philipp-spiess commented on packages/@tailwindcss-upgrade/src/index.ts:
```diff
@@ -0,0 +1,74 @@
+
+    files = await fastGlob(['**/*.css'], {
+      absolute: true,
+      ignore: ['**/node_modules', '**/vendor'],
```
Wonder if we should instead use the `.gitignore` info here, I remember that we had a few people who had `vendor` folders when we added the content-path-warning and we eventually had to revert it.

RobinMalfait replied:
Hmm yeah, let's see what I can find to solve this!

RobinMalfait replied:
Found globby https://github.com/sindresorhus/globby, based on fast-glob and has support for this out of the box

philipp-spiess replied:
:chef-kiss:

philipp-spiess commented on integrations/cli/codemods.test.ts:
Should we call this file `upgrade.test.ts` so it's clear that this is for the upgrade CLI?

RobinMalfait replied:
I think that might be better yeah, good catch!

RobinMalfait committed:
rename `codemods.test.ts` -> `upgrade.test.ts`

RobinMalfait committed:
use globby instead of fast-glob

Globby, created by Sindresorhus, is based on fast-glob but with
additional features. One of those features is the `gitignore` feature
which respects your `.gitignore` file, which is exactly what we need.

See: https://github.com/sindresorhus/globby#gitignore

RobinMalfait committed:
update changelog
